router-tests/headers_test.go

```diff
@@ -227,3 +228,230 @@ func TestForwardHeaders(t *testing.T) {
 		})
 	})
 }
+
+func TestForwardRenamedHeaders(t *testing.T) {
+	t.Parallel()
+
+	const (
+		// Make sure you copy these to the struct tag in the subscription test
+		headerNameInGlobalRule     = "foo"
+		headerNameInSubgraphRule   = "barista" // This matches the regex in test1 subgraph forwarding rules
+		headerRenameInGlobalRule   = "light"
+		headerRenameInSubgraphRule = "cell" // This matches the regex in test1 subgraph forwarding rules
+		headerValue                = "bar"
+		headerValue2               = "baz"
+
+		subscriptionForGlobalRulePayload   = `{"query": "subscription { headerValue(name:\"light\", repeat:3) { value }}"}`
+		subscriptionForSubgraphRulePayload = `{"query": "subscription { headerValue(name:\"cell\", repeat:3) { value }}"}`
+	)
+
+	headerRules := config.HeaderRules{
+		All: config.GlobalHeaderRule{
+			Request: []config.RequestHeaderRule{
+				{
+					Operation: config.HeaderRuleOperationPropagate,
+					Named:     headerNameInGlobalRule,
+					Rename:    headerRenameInGlobalRule,
+				},
+			},
+		},
+		Subgraphs: map[string]config.GlobalHeaderRule{
+			"test1": {
+				Request: []config.RequestHeaderRule{
+					{
+						Operation: config.HeaderRuleOperationPropagate,
+						Matching:  "(?i)^bar.*",
+						Rename:    headerRenameInSubgraphRule,
+					},
+				},
+			},
+		},
+	}
+
+	t.Run("HTTP", func(t *testing.T) {
+		t.Parallel()
+
+		cases := []struct {
+			headerName   string
+			headerRename string
+			testName     string
+		}{
+			{headerNameInGlobalRule, headerRenameInGlobalRule, "global rule"},
+			{headerNameInSubgraphRule, headerRenameInSubgraphRule, "subgraph rule"},
+		}
+		testenv.Run(t, &testenv.Config{
+			ModifyEngineExecutionConfiguration: func(cfg *config.EngineExecutionConfiguration) {
+				cfg.WebSocketReadTimeout = time.Millisecond * 10
+			},
+			RouterOptions: []core.Option{
+				core.WithHeaderRules(headerRules),
+			},
+		}, func(t *testing.T, xEnv *testenv.Environment) {
+			for _, c := range cases {
+				headerName := c.headerName
+				log.Println(c.testName, c.headerName, c.headerRename)
+				log.Println(headerName, headerValue)
+				t.Run(c.testName, func(t *testing.T) {
+					res := xEnv.MakeGraphQLRequestOK(testenv.GraphQLRequest{
+						Header: http.Header{
+							headerName: []string{headerValue},
+						},
+						Query: `query { headerValue(name:"` + c.headerRename + `") }`,
+					})
+					log.Println(res.Body)
+					require.Equal(t, `{"data":{"headerValue":"`+headerValue+`"}}`, res.Body)
+
+				})
+			}
+		})
+	})
+
+	t.Run("HTTP with client extension", func(t *testing.T) {
```

Comment:
What does it test regarding header renaming? You could query `headerValue` twice.